src/Tinfoil/Data/KDF.hs

```diff
@@ -64,7 +64,7 @@ data Verification = Verification !Verified !NeedsRehash
 --  processors (GPUs, mining ASICs, et cetera).
 data KDF = KDF
   { genHash        :: (Credential -> IO CredentialHash)
-  , hashCredential :: (Credential -> CredentialHash -> IO Verified)
-  , verifyNoHash   :: IO Verified
+  , hashCredential :: (CredentialHash -> Credential -> IO (Maybe Verified))
+  , verifyNoHash   :: (Credential -> IO (Maybe Verified))
```

Comment:
Sorry, misunderstood you - you mean `Unverified` instead of the `Nothing`, right? I considered that and decided against it because it removes any visibility for clients into issues like "password database is corrupted". I could go either way though, do you think just having `Unverified` is a better option?